src/main.c

```diff
@@ -233,27 +403,120 @@ static void write_spc(FILE *f) {
 		BYTE *new_spc = memcpy(malloc(spc_size), res_data, spc_size);
 
 		// Copy packs/blocks to byte array
-		for (int pack = 0; pack < 3; pack++) {
-			pack_to_spc(packs_loaded[pack], new_spc);
+//		for (int pack = 0; pack < 3; pack++) {
+//			pack_to_spc(packs_loaded[pack], new_spc);
+//		}
+
+		// compile_song corrupts the spc and any potential samples/instruments, so we need to make a copy first...
+		BYTE spc_copy[0x10000];
+		memcpy(spc_copy, spc, 0x10000);
+		struct SamplePointers { WORD start, loop; } *sample_pointers = (struct SamplePointers *)&spc_copy[sample_ptr_base];
+
+		// Move the music from wherever it was in the spc to 0x3200... (ripping out part of the program block in the process...)
+		const WORD dstMusic = 0x3200;
+		const int music_size = compile_song(&cur_song);
+		memcpy(spc + dstMusic, &spc[cur_song.address], music_size);
+
+		// recompile so the addresses are correct...
+		cur_song.address = dstMusic;
+		compile_song(&cur_song);
+
+		const unsigned int NUM_INSTRUMENTS = 80;
+		const WORD inst_size = NUM_INSTRUMENTS*6;
+		// Calculate buffer needed for sample pointer table to round to nearest 0x100.
+		const WORD REMAINDER = 0x100 - ((dstMusic + music_size) & 0xFF);
+		const WORD dstSamplePointers = dstMusic + music_size + REMAINDER;
+		const WORD BUFFER = 0x10; // Generic buffer between data...
+		const WORD dstInstruments = dstSamplePointers + BUFFER + 0x2*NUM_INSTRUMENTS;
```

Comment:
I'm having trouble following this. What does `BUFFER` actually do? And shouldn't this be 4 bytes per instrument, for the sample pointers? (start is two bytes, loop is two bytes)

Reply:
Also, yes, that should've been 4 bytes, not two. That was a mistake that was originally made but never corrected. Good catch